peloton/update_project.py

```diff
@@ -0,0 +1,1371 @@
+        mutation: sgqlc.operation.Operation,
+        index: int
+    ) -> None:
+        mutation.clear_project_v2_item_field_value(
+            input=self._inputs[index], __alias__=self._aliases[index]
+        ).client_mutation_id()
+
+
+class RemoveItemsMutation(PaginatedMutation):
+    """Mutation to remove items from the Peloton GitHub project."""
+    def _make_inputs(
+        self,
+        data_frame: pd.DataFrame,
+        project_cols: ProjectItemsQuery.ColNames,
+        issue_cols: IssuesQuery.ColNames,
+    ) -> list[dict]:
+        return [
+            # Expecting a DataFrame where the required column is the index.
+            (self._inputs_default | dict(item_id=row.name))
+            for _, row in data_frame.iterrows()
+        ]
+
+    def _add_selector(
+        self,
+        mutation: sgqlc.operation.Operation,
+        index: int
+    ) -> None:
+        mutation.delete_project_v2_item(
+            input=self._inputs[index], __alias__=self._aliases[index]
+        ).client_mutation_id()
+
+
+class UpdateItemsMutation(PaginatedMutation):
+    """
+    Mutation to update the fields of items in the Peloton project.
+
+    More complex than the other subclasses, which each did one 'thing' to one
+    issue/item/discussion. Here we need to update each field of each project
+    item.
+    """
+    def _make_inputs(
+        self,
+        data_frame: pd.DataFrame,
+        project_cols: ProjectItemsQuery.ColNames,
+        issue_cols: IssuesQuery.ColNames,
+    ) -> list[dict]:
+        inputs = []
+
+        if data_frame.empty:
+            # The loop below is not going to run, so we don't need a query to
+            #  populate field_types.
+            field_types = None
+        else:
+            project_fields = ProjectFieldsQuery()
+            field_types = project_fields.data_frame[project_fields.cols.data_type]
+
+        def value_kwarg(field_id_: str, value) -> dict:
+            """
+            Use the data types in the GH schema to cast each value correctly.
+            """
+            if field_types is not None:
+                field_type_ = field_types.loc[field_id_]
+                if field_type_ is github_schema.ProjectV2FieldValue.date:
+                    # Convert a date-time string to a date string.
+                    value = value[:10]
+                return {field_type_.name: field_type_.type(value)}
+            else:
+                message = (
+                    "Problem with updating project item fields: "
+                    "cannot access information on field data types."
+                )
+                raise ValueError(message)
+
+        for _, row in data_frame.iterrows():
+            kwargs = (
+                self._inputs_default
+                |
+                dict(item_id=row[project_cols.id_in_project])
+            )
+
+            # The field in the project that contains the ID of the linked item.
+            link_field = "PVTF_lADOABU7f84ALhAIzgPKqWg"
```

Comment:
@trexfeathers Could this be a class member constant?

Reply:
65ed1b36807721b629547d9115ddc7f4f2b1e124